### PR TITLE
[BIOMAGE-1383] UI shows old backend status state before rerunning gem2s.

### PR DESCRIPTION
**Pull request by StefanBabukov**
# Background
#### Link to issue 
https://biomage.atlassian.net/secure/RapidBoard.jspa?rapidView=1&projectKey=BIOMAGE&modal=detail&selectedIssue=BIOMAGE-1383
#### Anything else the reviewers should know about the changes here

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [ ] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [ ] Deployed to staging
- [ ] Passed integration tests 

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

**Comment by aerlaut** [comment]
This might be fixed by #499

**Comment by StefanBabukov** [comment]
Yeah i branched off from that PR. There were further changes needed though:
API: before we checked if the gem2s parameters are changed when we run createGem2s with `shouldRunGem2s`. I moved that to the pipeline-status endpoint so when we get the backend status we may return "NEEDS_RERUN", which triggers gem2s.

UI: Added a `deleteBackendStatus` reducer to delete the status when you remove a project. Further changes are to where `runGem2s` is called - we need to call it in `ContentWrapper` with a condition based on the status. So now we don't run gem2s everytime and check if it needs to be run in the API, but we do that in the UI based on the backend status.

File: src/redux/actions/projects/deleteProject.js [hunk 1]
```diff
@@ -71,6 +71,10 @@ const deleteProject = (
       },
     });
 
+    dispatch({
+      type: BACKEND_STATUS_DELETED,
```
Comment (cosa65) [inline src/redux/actions/projects/deleteProject.js]: You're only using `BACKEND_STATUS_DELETED` to react to an experiment getting deleted, right? I think you could reuse the `EXPERIMENTS_DELETED` action, look into this: https://redux.js.org/style-guide/style-guide#allow-many-reducers-to-respond-to-the-same-action.

Basically you can just add in backend status a reducer for EXPERIMENTS_DELETED, that way you can get rid of `BACKEND_STATUS_DELETED`.

There is a good example of this in `EXPERIMENT_SETTINGS_PIPELINE_START`, you can see it is handled across a lot of small reducers. 

Ideally we should move away from dispatching many actions sequentially for one action creator. I think the problem is that we were creating and using redux actions to modify the store while we should instead have think of the store as reacting as a response to actions.

File: src/redux/actions/pipeline/runGem2s.js [hunk 2]
```diff
@@ -7,24 +8,26 @@ import {
...
+  const runningStatus = {
+    ...backendStatus[experimentId].status,
+    gem2s: {
+      status: pipelineStatus.RUNNING,
+      completedSteps: [],
+    },
+  };
   dispatch({
-    type: BACKEND_STATUS_LOADING,
+    type: BACKEND_STATUS_LOADED,
```
Comment (cosa65) [inline src/redux/actions/pipeline/runGem2s.js:25]: Similar thing here, we could have the backend status reducer react to EXPERIMENT_SETTINGS_PIPELINE_START (or EXPERIMENT_SETTINGS_GEM2S_START as a new action, I am not sure what is the best option here) and as a consequence set itself to the status you want in the reducer.

Moving this logic to a reducer will also help with this https://redux.js.org/style-guide/style-guide#put-as-much-logic-as-possible-in-reducers

File: src/components/data-management/AnalysisModal.jsx [hunk 3]
```diff
@@ -48,15 +44,13 @@ const AnalysisModal = (props) => {
...
     onLaunch(experimentId);
-    const analysisPath = '/experiments/[experimentId]/data-processing';
     const lastViewed = moment().toISOString();
     dispatch(updateExperiment(experimentId, { lastViewed }));
     dispatch(updateProject(activeProjectUuid, { lastAnalyzed: lastViewed }));
-
-    dispatch(runGem2s(experimentId));
-    router.push(analysisPath.replace('[experimentId]', experimentId));
+    const analysisPath = '/experiments/[experimentId]/data-processing';
+    window.location.href = await analysisPath.replace('[experimentId]', experimentId);
```
Comment (cosa65) [inline src/components/data-management/AnalysisModal.jsx:53]: Why use `window.location.href` and make it async instead of using the router?

File: src/components/GEM2SLoadingScreen.jsx [hunk 4]
```diff
@@ -46,6 +46,14 @@ const GEM2SLoadingScreen = (props) => {
       image: '/undraw_Dev_focus_re_6iwt.svg',
       alt: 'A woman working in front of a computer.',
     },
+    needsRerun: {
```
Comment (cosa65) [inline src/components/GEM2SLoadingScreen.jsx]: I couldn't find where this is being used, was it missing?

File: src/components/ContentWrapper.jsx [hunk 5]
```diff
@@ -96,15 +86,12 @@ const ContentWrapper = (props) => {
 
       io.on(`ExperimentUpdates-${experimentId}`, (update) => cb(experimentId, update));
     })();
-  }, [experimentId]);
-
-  useEffect(() => {
-    if (backendStatusRequested) {
-      return;
+    if ((![RUNNING, SUCCEEDED].includes(gem2sStatusKey)
```
Comment (cosa65) [inline src/components/ContentWrapper.jsx:89]: Checking whether we can re run or not gem2s or the qc pipeline is handled in the api, I don't think we need to add more checks here too, just to keep it simple.

File: src/components/ContentWrapper.jsx [hunk 6]
```diff
@@ -51,41 +50,32 @@ const ContentWrapper = (props) => {
...
 
   const changedQCFilters = useSelector(
     (state) => state.experimentSettings.processing.meta.changedQCFilters,
   );
 
-  // This is used to prevent a race condition where the page would start loading immediately
-  // when the backend status was previously loaded. In that case, `backendLoading` is `false`
-  // and would be set to true only in the `loadBackendStatus` action, the time between the
-  // two events would allow pages to load.
-  const [backendStatusRequested, setBackendStatusRequested] = useState(false);
```
Comment (cosa65) [inline src/components/ContentWrapper.jsx:79]: Why was this removed?

File: src/redux/reducers/backendStatus/index.js [hunk 7]
```diff
@@ -21,6 +25,18 @@ const backendStatusReducer = (state = initialState, action) => {
     case BACKEND_STATUS_ERROR: {
       return backendStatusError(state, action);
     }
+    case BACKEND_STATUS_DELETED: {
+      return backendStatusDeleted(state, action);
+    }
+    case HYDRATE: {
```
Comment (cosa65) [inline src/redux/reducers/backendStatus/index.js:30]: Out of ignorance, what is this used for?

File: src/utils/pipelineStatusValues.js [hunk 8]
```diff
@@ -8,7 +8,7 @@ export default {
   TIMED_OUT: 'TIMED_OUT',
   ABORTED: 'ABORTED',
   SUCCEEDED: 'SUCCEEDED',
-
+  NEEDS_RERUN: 'NEEDS_RERUN',
```
Comment (cosa65) [inline src/utils/pipelineStatusValues.js:11]: We won't need this if we remove the checks for gem2s in the ui, right?

File: src/redux/actions/backendStatus/loadBackendStatus.js [hunk 9]
```diff
@@ -17,7 +17,9 @@ const loadBackendStatus = (experimentId) => async (dispatch) => {
 
   const url = `/v1/experiments/${experimentId}/backendStatus`;
   try {
-    const response = await fetchAPI(url);
+    const response = await fetchAPI(url, {}, {
+      uiUrl,
```
Comment (cosa65) [inline src/redux/actions/backendStatus/loadBackendStatus.js:21]: What is this being used for?

File: src/redux/actions/projects/deleteProject.js (hunk 1 above)
Comment (StefanBabukov) [reply src/redux/actions/projects/deleteProject.js:75]: Yeah I thought of that, but in the future we will have many experiments associated to a project right? So in this case we might need to use this backend_status_deleted in other places (when you delete a particullar experiment, not the entire project)

File: src/components/data-management/AnalysisModal.jsx (hunk 3 above)
Comment (StefanBabukov) [reply src/components/data-management/AnalysisModal.jsx:53]: these are changes introduced by Marcell from this PR -https://github.com/biomage-ltd/ui/pull/499

File: src/components/ContentWrapper.jsx (hunk 5 above)
Comment (StefanBabukov) [reply src/components/ContentWrapper.jsx:89]: I removed those checks because its much faster and i think better to only call the api when needed. If the checks are done in the api it is delayed a little bit so thats why the user is in data-processing first and then gem2s is triggered again, what do you think?

File: src/components/ContentWrapper.jsx (hunk 6 above)
Comment (StefanBabukov) [reply src/components/ContentWrapper.jsx:79]: these are changes introduced by Marcell from this PR -https://github.com/biomage-ltd/ui/pull/499

File: src/redux/reducers/backendStatus/index.js (hunk 7 above)
Comment (StefanBabukov) [reply src/redux/reducers/backendStatus/index.js:30]: these are changes introduced by Marcell from this PR -https://github.com/biomage-ltd/ui/pull/499 i believe its for when the SSR happens to update the status

File: src/redux/actions/backendStatus/loadBackendStatus.js (hunk 9 above)
Comment (StefanBabukov) [reply src/redux/actions/backendStatus/loadBackendStatus.js:21]: these are changes introduced by Marcell from this PR -https://github.com/biomage-ltd/ui/pull/499

File: src/redux/actions/projects/deleteProject.js (hunk 1 above)
Comment (StefanBabukov) [reply src/redux/actions/projects/deleteProject.js:75]: What do you think?

File: src/utils/pipelineStatusValues.js (hunk 8 above)
Comment (StefanBabukov) [reply src/utils/pipelineStatusValues.js:11]: Maybe i can remove the checks and pass the status to the api? So the checks are done there?

File: src/redux/actions/backendStatus/loadBackendStatus.js (hunk 9 above)
Comment (ivababukova) [reply src/redux/actions/backendStatus/loadBackendStatus.js:21]: still, do figure out what this is for and let us know

File: src/components/ContentWrapper.jsx (hunk 6 above)
Comment (ivababukova) [reply src/components/ContentWrapper.jsx:79]: still, do work out what it is for and why it is added. Maybe Marcell mistakenly removed it? We cannot be merging in things in the code that we don't understand, because we risk breaking something.

File: src/components/data-management/AnalysisModal.jsx (hunk 3 above)
Comment (ivababukova) [reply src/components/data-management/AnalysisModal.jsx:53]: again, do figure that out

**Review by ivababukova — CHANGES_REQUESTED** [review]
Code coverage is falling here. We need to add unit tests for what is changed, but not covered

**Comment by StefanBabukov** [comment]
I'll add them once im settled with the final changes

File: src/components/ContentWrapper.jsx (hunk 5 above)
Comment (cosa65) [reply src/components/ContentWrapper.jsx:89]: @kafkasl what do you think about this? I am not sure whether the check for whether gem2s should be rerun should be done in the api or the ui

**Comment by ivababukova** [comment]
Closing this PR, as these changes are getting implemented as part of Agi's ticket
